### PR TITLE
fix: fix explore table filtering and sorting bugs

### PR DESCRIPTION
**Pull request by lynnshaoyu**
1) get rid of data flickering when you change time period 
2) fix search string being ignored if you change the time period you’re viewing
3) make search filter queries reset /disappear when you navigate to a different page and come back. 
4) when user starts typing a search term, show correct number of loading rows (we have this info from prefetched top 100 tokens query), and load actual rows once price history token data becomes available. 
5) add local "cache" of preloaded tokens with price history (if we have already rendered them previously)

### CODE REVIEW

**Comment by JFrankfurt** [comment]
typed in a search string to filter the results:
- the two resulting list elements flipped to a loading state (I don't think they should have, since they were already rendered)
- two POST requests were sent, but the second was empty

**Comment by JFrankfurt** [comment]
nit: if I text filter for a token here's what I experience:
- full page loading experience (expected)
- loading experience for the rows that result from my search string (weird! we know enough to show those rows, so why are they in a loading state??)
- loaded rows render 👍

It seems like it should be possible to drop the second state from this flow!

**Comment by JFrankfurt** [comment]
not sure how, but I was able to get to a state where it seemed like my filter string was ignored

**Comment by lynnshaoyu** [comment]
> not sure how, but I was able to get to a state where it seemed like my filter string was ignored

ooof can u lmk how to repro if you figure out the steps :O

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 1]
```diff
@@ -55,13 +54,13 @@ function NoTokensState({ message }: { message: ReactNode }) {
...
-const InitialLoadingRows = Array(ROWS_PER_PAGE_FETCH).fill(<LoadingRow />)
+const LoadingRows = (rowCount?: number) => Array(rowCount ?? PAGE_SIZE).fill(<LoadingRow />)
 
-export function LoadingTokenTable() {
+export function LoadingTokenTable({ rowCount }: { rowCount?: number }) {
   return (
     <GridContainer>
       <HeaderRow />
-      <TokenDataContainer>{InitialLoadingRows}</TokenDataContainer>
+      <TokenDataContainer>{LoadingRows(rowCount)}</TokenDataContainer>
```
Comment (JFrankfurt) [inline src/components/Tokens/TokenTable/TokenTable.tsx:63]: <3

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 2]
```diff
@@ -118,14 +115,14 @@ export default function TokenTable() {
             <TokenDataContainer>
               {tokens.map((token, index) => (
                 <LoadedRow
-                  key={token?.name}
+                  key={token?.address}
                   tokenListIndex={index}
                   tokenListLength={tokens?.length ?? 0}
                   token={token}
-                  ref={tokens.length === index + 1 ? lastTokenRef : undefined}
+                  ref={index + 1 === tokens.length ? lastTokenRef : undefined}
```
Comment (JFrankfurt) [inline src/components/Tokens/TokenTable/TokenTable.tsx:122]: why this change?

File: src/components/Tokens/TokenTable/TokenTable.tsx [hunk 3]
```diff
@@ -118,14 +115,14 @@ export default function TokenTable() {
...
                 />
               ))}
-              {loading && LoadingMoreRows}
```
Comment (JFrankfurt) [inline src/components/Tokens/TokenTable/TokenTable.tsx:128]: compress the compound boolean render conditional to a computed single boolean:
`const showMoreLoadingRows = Boolean(loading && hasMore)`
...
`{showMoreLoadingRows && LoadingMoreRows}`

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 2 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenTable/TokenTable.tsx:122]: lmao because my brain thinks this makes more sense XD 

we are checking if the current index is equal to the tokens length, not vice versa. there is no diff tbh.

File: src/graphql/data/TopTokens.ts [hunk 4]
```diff
@@ -129,63 +130,101 @@ function toContractInput(token: PrefetchedTopToken) {
...
-  loadMoreTokens: () => void
+// Map of key: ${chain} + ${address} and value: TopToken object.
+// Acts as a local cache.
+const tokensWithPriceHistoryCache: Record<string, TopToken> = {}
+
+const checkIfAllTokensCached = (prefetchedSelectedTokens: PrefetchedTopToken[]) => {
+  let everyTokenHasBeenPreloaded = true
+  const preloadedSelectedTokens: TopToken[] = []
+  prefetchedSelectedTokens.map((token) =>
+    !!token && `${token.chain.toString()}${token.address}` in tokensWithPriceHistoryCache
```
Comment (JFrankfurt) [inline src/graphql/data/TopTokens.ts:141]: I think `.toString` will be called here implicitly, so you don't need to include it.

File: src/graphql/data/TopTokens.ts [hunk 5]
```diff
@@ -129,63 +130,101 @@ function toContractInput(token: PrefetchedTopToken) {
...
-      complete() {
+  const loadMoreTokens = useCallback(() => {
+    setLoading(true)
+    const contracts = prefetchedSelectedTokensWithoutPriceHistory
+      .slice(page * PAGE_SIZE, (page + 1) * PAGE_SIZE)
+      .map(toContractInput)
+    loadTokensWithPriceHistory(contracts, (data) => {
+      if (data?.tokens) {
+        data.tokens.forEach((token) =>
+          !!token ? (tokensWithPriceHistoryCache[`${token.chain.toString()}${token.address}`] = token) : null
```
Comment (JFrankfurt) [inline src/graphql/data/TopTokens.ts:192]: same comment here re: `.toString`

File: src/graphql/data/TopTokens.ts [hunk 6]
```diff
@@ -129,63 +130,101 @@ function toContractInput(token: PrefetchedTopToken) {
...
+      setLoading(false)
+      return
+    } else {
+      setLoading(true)
+      setTokens([])
+      const contracts = prefetchedSelectedTokensWithoutPriceHistory.slice(0, PAGE_SIZE).map(toContractInput)
+      loadTokensWithPriceHistory(contracts, (data) => {
+        if (data?.tokens) {
+          data.tokens.forEach((token) =>
+            !!token ? (tokensWithPriceHistoryCache[`${token.chain.toString()}${token.address}`] = token) : null
```
Comment (JFrankfurt) [inline src/graphql/data/TopTokens.ts:217]: same here re: `.toString`

**Review by JFrankfurt — CHANGES_REQUESTED** [review]
see comments

File: src/graphql/data/TopTokens.ts (hunk 5 above)
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts:192]: done

File: src/graphql/data/TopTokens.ts (hunk 4 above)
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts:141]: done

File: src/graphql/data/TopTokens.ts (hunk 6 above)
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts:217]: done

File: src/graphql/data/TopTokens.ts [hunk 7]
```diff
@@ -120,7 +121,7 @@ function useFilteredTokens(tokens: PrefetchedTopToken[]) {
   }, [tokens, showFavorites, lowercaseFilterString, favorites])
 }
 
-const PAGE_SIZE = 20
+export const PAGE_SIZE = 20
```
[the remark this answers is not in the document]
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts:124]: done

File: src/components/Tokens/TokenTable/SearchBar.tsx [hunk 8]
```diff
@@ -59,10 +59,15 @@ const SearchInput = styled.input`
...
 export default function SearchBar() {
-  const [localFilterString, setLocalFilterString] = useState('')
+  const currentString = useAtomValue(filterStringAtom)
+  const [localFilterString, setLocalFilterString] = useState(currentString)
   const setFilterString = useUpdateAtom(filterStringAtom)
   const debouncedLocalFilterString = useDebounce(localFilterString, 300)
 
+  useEffect(() => {
+    setLocalFilterString(currentString)
+  }, [currentString])
```
Comment (cmcewen) [inline src/components/Tokens/TokenTable/SearchBar.tsx:69]: where else in the app would be updating this value?

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 1 above)
Comment (cmcewen) [reply src/components/Tokens/TokenTable/TokenTable.tsx:63]: do these rows need keys?

File: src/graphql/data/TopTokens.ts [hunk 9]
```diff
@@ -8,18 +8,19 @@ import {
...
+import { useCurrentChainName } from './util'
 
-export function usePrefetchTopTokens() {
-  const duration = toHistoryDuration(useAtomValue(filterTimeAtom))
+export function usePrefetchTopTokens(duration: HistoryDuration) {
   const chain = useCurrentChainName()
-  const args = useMemo(() => ({ chain, duration }), [chain, duration])
-  return useLazyLoadQuery<TopTokens100Query>(topTokens100Query, args)
+
+  const top100 = useLazyLoadQuery<TopTokens100Query>(topTokens100Query, { duration, chain })
```
Comment (cmcewen) [inline src/graphql/data/TopTokens.ts:22]: why not just return this value instead of assigning and returning that?

File: src/graphql/data/TopTokens.ts [hunk 10]
```diff
@@ -129,63 +130,101 @@ function toContractInput(token: PrefetchedTopToken) {
...
+const tokensWithPriceHistoryCache: Record<string, TopToken> = {}
+
+const checkIfAllTokensCached = (prefetchedSelectedTokens: PrefetchedTopToken[]) => {
+  let everyTokenHasBeenPreloaded = true
+  const preloadedSelectedTokens: TopToken[] = []
+  prefetchedSelectedTokens.map((token) =>
+    !!token && `${token.chain.toString()}${token.address}` in tokensWithPriceHistoryCache
+      ? preloadedSelectedTokens.push(tokensWithPriceHistoryCache[`${token.chain.toString()}${token.address}`])
+      : (everyTokenHasBeenPreloaded = false)
+  )
```
[the remark this answers is not in the document]
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts:144]: sounds good, made it clearer.

File: src/graphql/data/TopTokens.ts [hunk 11]
```diff
@@ -129,63 +130,108 @@ function toContractInput(token: PrefetchedTopToken) {
...
+  }, [prefetchedSelectedTokensWithoutPriceHistory, page, loadTokensWithPriceHistory, tokens])
 
-  return { loading, tokens: useFilteredTokens(useSortedTokens(tokens)) as TopToken[], loadMoreTokens }
+  // Reset count when filters are changed
+  useLayoutEffect(() => {
+    const { everyTokenHasBeenPreloaded, preloadedSelectedTokens } = checkIfAllTokensCached(
+      prefetchedSelectedTokensWithoutPriceHistory
+    )
+    if (everyTokenHasBeenPreloaded) {
+      setTokens([...preloadedSelectedTokens])
```
Comment (cmcewen) [inline src/graphql/data/TopTokens.ts:207]: why do you need to spread here?

File: src/graphql/data/TopTokens.ts [hunk 12]
```diff
@@ -129,63 +130,108 @@ function toContractInput(token: PrefetchedTopToken) {
...
+      setLoading(false)
+      return
+    } else {
+      setLoading(true)
+      setTokens([])
+      const contracts = prefetchedSelectedTokensWithoutPriceHistory.slice(0, PAGE_SIZE).map(toContractInput)
+      loadTokensWithPriceHistory(contracts, (data) => {
+        if (data?.tokens) {
+          data.tokens.forEach((token) =>
```
Comment (cmcewen) [inline src/graphql/data/TopTokens.ts:216]: why do you mutate the array vs map here? would that fix the TS error?

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 3 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenTable/TokenTable.tsx:128]: done

File: src/graphql/data/TopTokens.ts (hunk 9 above)
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts:22]: that is a good idea haha, fixed.

File: src/components/Tokens/TokenTable/SearchBar.tsx (hunk 8 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenTable/SearchBar.tsx:69]: https://github.com/Uniswap/interface/blob/0b9fda5b25adc56fed0a29ca84eb7d38c62ca348/src/pages/Tokens/index.tsx#L71

File: src/components/Tokens/TokenTable/TokenTable.tsx (hunk 1 above)
Comment (lynnshaoyu) [reply src/components/Tokens/TokenTable/TokenTable.tsx:63]: dont think so they're all the same

File: src/graphql/data/TopTokens.ts (hunk 12 above)
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts]: good idea, changed to map. that doesn't fix the ts error, but i spread the array instead and it fixed it

File: src/graphql/data/TopTokens.ts (hunk 11 above)
Comment (lynnshaoyu) [reply src/graphql/data/TopTokens.ts]: i dont need to :P removed spread.

File: src/components/Tokens/TokenTable/SearchBar.tsx (hunk 8 above)
Comment (cmcewen) [reply src/components/Tokens/TokenTable/SearchBar.tsx:69]: makes sense 👍

**Review by cmcewen — APPROVED** [review]
